pytorch_lightning/utilities/fetching.py

```diff
@@ -0,0 +1,140 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+from collections.abc import Iterable, Iterator
+from typing import Any, Generator, List, Optional, Tuple
+
+from torch.utils.data.dataloader import DataLoader
+
+from pytorch_lightning.trainer.supporters import CombinedLoader
+from pytorch_lightning.utilities.apply_func import apply_to_collection
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+
+class AbstractFetcher(ABC):
+
+    """
+    This class is used to control batch fetching flow.
+    """
+
+    @abstractmethod
+    def fetching_function(self) -> Generator:
+        pass
+
+    def __init__(
+        self,
+        prefetch_batches: int = 1,
```

Comment:
does `prefetch_batches=1` mean that there is one additional batch pre-fetched (while the current one is processed) or is this the current behaviour with sequential fetching? If it is the current, I think this is ambigious and it should rather be 0 for this behaviour.

I would strongly advocate for defaulting to the current behaviour and making this opt-in always since there are many scenarios, where only a single batch fits into memory.

Reply:
Done !

Reply:
@justusschock I made the update, but I am not sure to like it. I mean this isn't final yet.
@awaelchli @ananthsub @carmocca Any thoughts ?